navapbc/template-application-nextjs
# Proof-of-concept: Support OAuth and OIDC auth providers

> [!WARNING]  
> **This is just a proof-of-concept. There are a number of TODO's that would need cleaned up before merging.** 

## Ticket

Resolves #235

## Changes

- Enable Next.js app router
- Add `next-auth` with an example Cognito integration

## Initial impressions of Auth.js

- Using Auth.js to provide OAuth provider support seems like a beneficial addition to the template, and would support a project team when creating an MVP authentication integration. 
- It's not limited to integrating with Cognito — **it would support any OAuth/OIDC provider, including Login.gov**.
- By default, the session is persisted to a JWT-encoded cookie, but it can be [configured to persist to a database instead](https://authjs.dev/getting-started/adapters).
- Auth.js seems great for projects that are utilizing the identity provider's hosted authentication flow (like Login.gov). If a team were building a fully custom flow for signup, password reset, etc, then would likely be better off with the identity provider's SDK instead.
- Some things are specific to the provider, which teams would be responsible for customizing in Auth.js, like [token refresh](https://authjs.dev/guides/basics/refresh-token-rotation) and federated sign-out. Token refresh is likely the thorniest and I could see the template expanding its support for this in the future, but I think it could be left to project teams initially.
- An end-to-end test suite is likely the best way to provide test coverage for most of this functionality. The template doesn't currently provide end-to-end testing functionality.

### Open questions

- Would we keep this Cognito implementation in the template to provide an example, and teams could swap it out if they're using a different provider?

## Testing

To test this locally:

1. Set the environment variables in `app/.env.local`. You can copy the values from 1password, see "Next.js platform env.local (next-auth)"
2. Visit http://localhost:3000/user

https://github.com/navapbc/template-application-nextjs/assets/371943/4818597f-df22-4ea3-bf3e-a38c4726ca31
